pkg/apis/logicalcluster/v1alpha1/cluster-provider-desc.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2022 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ClusterProviderType string
+
+const (
+	KindProviderType ClusterProviderType = "kind"
+)
+
+// ClusterProviderDesc represents a provider.
+//
+// +crd
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:resource:scope=Cluster,shortName=cpc
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ClusterProviderDesc struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// `spec` describes a provider.
+	// +optional
+	Spec ClusterProviderDescSpec `json:"spec,omitempty"`
+
+	// `status` describes the status of the cluster object.
+	// +optional
+	Status ClusterProviderDescStatus `json:"status"`
+}
+
+// ClusterProviderDescSpec describes a cluster.
+// TODO: We are currently only listing the type and config.
+// There will be additional fields in the future.
+type ClusterProviderDescSpec struct {
+	// ProviderType is the type of the provider of the cluster.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="cluster is immutable"
+	ProviderType ClusterProviderType `json:"ProviderType"`
+
+	// TODO: this should be stored as a secret!
```

Comment:
Should the whole config be in a Secret or can some of it be in the open?

Reply:
Yes, I assumed that different providers will need different info.  For example, a provider of kind clusters on the same machine needs no info at all.  Indeed some providers will likely need information that needs to be kept private.  So the simplest adequate solution would be to keep all of the needed information (if there is any at all), for a given provider, in one Secret object.  A possible further step would be to remove from that Secret object information that does not need to be kept private and instead hold it in a ConfigMap or something.

Reply:
As mentioned above, in stage 2, we will switch to REST based providers where the client connects through regular TLS based connection and the type of info needed to connect to a provider will be identical for all providers (e.g., crt, key, IP)